compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -4027,6 +4033,124 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
             }
         }
     }
+
+    fn suggest_desugaring_async_fn_in_trait(
+        &self,
+        err: &mut Diagnostic,
+        trait_ref: ty::PolyTraitRef<'tcx>,
+    ) {
+        // Don't suggest if RTN is active -- we should prefer a where-clause bound instead.
+        if self.tcx.features().return_type_notation {
+            return;
+        }
+
+        let trait_def_id = trait_ref.def_id();
+
+        // Only suggest specifying auto traits
+        if !self.tcx.trait_is_auto(trait_def_id) {
+            return;
+        }
+
+        // Look for an RPITIT
+        let ty::Alias(ty::Projection, alias_ty) = trait_ref.self_ty().skip_binder().kind() else {
+            return;
+        };
+        let Some(ty::ImplTraitInTraitData::Trait { fn_def_id, opaque_def_id }) =
+            self.tcx.opt_rpitit_info(alias_ty.def_id)
+        else {
+            return;
+        };
+
+        // ... which is a local function
+        let Some(fn_def_id) = fn_def_id.as_local() else {
+            return;
+        };
```

Comment:
Should we maybe *mention* that the problem lies in the trait definition so they know they have to bother the crate maintainer? Particularly thinking of the case of workspaces where the crate isn't *local*, but still under the dev's control.

Reply:
Yeah, I can do that. I can at least point to the async fn definition if we have a span for that.